### DISCUSSION TITLE
1-n relation mapping

### DISCUSSION THREAD

**Discussion by spidertwin2** [prisma/prisma#7346] [Q&A] [answered]
Hello,

My schema is as follows: 

User -- Profile : 1-1 relation
Profile -- Location: 1-n relation (a profile can only have 1 location, but a Location can be assigned to many profiles)

In the Location model field `profiles Profile[ ]` should I save all 5 fields of Profile, or is it better to just store an array of profile ids?

```prisma
model User {
  id             Int            @id @default(autoincrement())
  createdAt      DateTime       @default(now())
  updatedAt      DateTime       @updatedAt
  email          String         @unique
  hashedPassword String?
  role           UserRole       @default(USER)
  profile        Profile?
  tokens         Token[]
  sessions       Session[]
}
```
```prisma
model Profile {
  id         Int       @id @default(autoincrement())
  name       String?
  ageGroup   String?
  location   Location? @relation(fields: [locationId], references: [id])
  locationId Int?
  user       User      @relation(fields: [userId], references: [id])
  userId     Int       @unique
}
```

```prisma
model Location {
  id       Int       @id @default(autoincrement())
  name     String?
  label    String?   @unique
  locality String?
  region   String?
  country  String?
  profiles Profile[]
}
```

Thanks :parasol_on_ground:

**Comment by ryands17** [comment, marked as answer]
@spidertwin2 👋 

In this case, I would suggest keeping the relation `profiles Profile[]` in the `Location` model so that it will be easier to reference via [`include`](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#include). 

This will help fetching the relations easily with Prisma's syntax and will make it easier to delete the `Location` if you would want to delete the `Profile` as well (cascading deletes). Let me know if you have any further queries!

**Reply by spidertwin2** [reply]
Thanks @ryands17

I am unable to correctly update the `profiles` field with `upsert`.

When the user saves a city, prisma should -
- fetch user profile
- search for city in Location table
- if city found, add user profile to `profiles`
- if city not found, create new record in the Location table and add user profile to `profiles`
- update user profile with new `locationId`

updateLocation.ts

```ts
type LocationProps = {
  name?: string
  label?: string
  locality?: string
  region?: string
  country?: string
}

export default async function updateLocation(location: LocationProps, ctx: Ctx) {
  ctx.session.$authorize

  if (!ctx.session.userId) return null
  try {
    // find user profile
    const profile = await db.profile.findFirst({
      where: {
        userId: ctx.session.userId,
      },
    })

    // update or create location
    const findOrCreateLocation = await db.location.upsert({
      where: {
        label: location?.label, 
      },
      update: {
        profiles: profile, //error
      },
      create: {
        name: location?.name,
        label: location?.label,
        locality: location?.locality,
        region: location?.locality,
        country: location?.country,
        profiles: [profile], //error
      },
    })

    // update user profile with new locationId
    const updatedProfile = await db.profile.update({
      where: {
        id: profile?.id,
      },
      data: {
        locationId: findOrCreateLocation.id,
      },
    })

    console.log("updated profile: ", updatedProfile)

    return findOrCreateLocation
  } catch (error) {
    console.log("Error while updating your location: ", error)
  }
}
```
:palm_tree:

**Reply by ryands17** [reply]
You are missing `connect`. This would be how you add it:

```ts
  const profile = await prisma.profile.findFirst({
    where: {
      userId: 1,
    },
    rejectOnNotFound: true,
  })

  // update or create location
  const findOrCreateLocation = await prisma.location.upsert({
    where: {
      label: location?.label,
    },
    update: {
      profiles: { create: profile },
    },
    create: {
      name: location?.name,
      label: location?.label,
      locality: location?.locality,
      region: location?.locality,
      country: location?.country,
      profiles: { create: profile },
    },
  })

  // update user profile with new locationId
  const updatedProfile = await prisma.profile.update({
    where: {
      id: profile.id,
    },
    data: {
      locationId: findOrCreateLocation.id,
    },
  })
```

You can read more about this [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#create-a-related-record)

Also the `LocationProps` can be re-written to make use of Prisma's typing as follows:

```ts
import { Location } from '@prisma/client'

type LocationProps = Omit<Location, 'id'>
```

This will make the typings similar to what the database expects.

**Reply by spidertwin2** [reply]
works now, thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
